keep track of the created wrappedFn

valdrinkoshi opened the pull request:
Fixes #24 by keeping track of the node, the event type and event options when patching add/removeEventListener.
Removed `__eventListenerCount` as it doesn't seem to be used anywhere (better perf).

@azakus @sorvell PTAL

valdrinkoshi commented on src/event-mixin.js:
```diff
@@ -263,10 +268,11 @@ export function removeEventListener(type, fn, optionsOrCapture) {
+  const wrapper = wrappersForType.pop();
   origRemoveEventListener.call(this, type, wrapper || fn, optionsOrCapture);
   if (wrapper) {
-    fn.__eventWrapper = null;
```
I've purposefully omitted the cleanup here, as it might be expensive to check if `fn.__eventWrappers` is actually empty. WDYT?

sorvell commented:
The implementation here isn't quite right. A call to `addEventListener` with the same arguments as a previous call will not add another listener. So we need to disambiguate between the element, type, and options (3rd argument) passed to `addEventListener`. The element/type could be mapped, but the options are not identity checked so this requires more bookkeeping. 

If that's clear enough @valdrinkoshi would you mind fixing that up?

valdrinkoshi commented:
@sorvell @azakus updated to take useCapture|options in consideration, added tests too 👌

valdrinkoshi commented:
err, this is still incorrect, as the following should invoke `callback` only once:
``js
document.body.addEventListener('click', callback);
document.body.addEventListener('click', callback);
document.body.click();
```
working on it..!

valdrinkoshi commented:
@sorvell PR ready for review, I've incorporated the feedback.
I've also added the support of the `once` option (the listener gets automatically removed when `once = true`).

sorvell commented on src/event-mixin.js:
```diff
@@ -226,7 +258,28 @@ export function addEventListener(type, fn, optionsOrCapture) {
+  // This event listener might be added multiple times, we need to be able to remove
+  // all the wrappers we add.
+  const eventWrapperId = getEventWrapperId(type, optionsOrCapture);
+  const wrappersForType = fn.__eventWrappers[eventWrapperId] || new WeakMap();
```
Have a performance concern here. Almost always there will be a new wrapper id when addEventListener is called and this means we'll basically make a new WeakMap for every call to addEventListener. It seems highly likely this will result in poor performance. At the least we must test this and I think it's probably better to refactor this to reduce the number of WeakMaps needed.

sorvell replied:
fwiw, I'd be more willing to make `removeEventListener` slow v. `addEventListener`.

sorvell replied:
Also note we already make an `__dom` object to track various info on the element and this could be used to help here as well.

sorvell replied:
If a more optimal approach isn't obvious, I'd be willing to take this PR with a comment added to keep an eye on the related perf.

valdrinkoshi replied:
challenge accepted! will be back with an answer soon 😁

valdrinkoshi replied:
@sorvell I've tried few options & used this benchmark to measure how slow this patch is: synchronously add 1000 different event listeners, measure how long it takes https://gist.github.com/valdrinkoshi/32d33fd66b7cbcd48a9f70f30fb572ce
The native addEventListener takes ~2-3ms, the current approach takes ~8-10ms (2.5-3x more)
I've tried not relying on the WeakMap by generating the eventWrapperId like this:
```js
let nodeIdCount = 0;
function getEventWrapperId(node, type, options) {
  node.__id = node.__id || node.id || ++nodeIdCount;
  return node.__id + '|' + type + '|' + normalizedOptions(options);
}
```
This allows to have only 1 big object `fn.__eventWrappers` that we fill with these keys, and have `fn._eventWrappers[wrapperId] = wrappedFn`.
Performance got worse (~12-14ms, 5-6x slower).

WDYT?

valdrinkoshi commented:
@sorvell now the approach is to track the wrapped function by storing the type/options/node/wrapperFn into the `fn.__eventWrappers` array, rather than storing type/options/fn/wrapperFn into the `node.__eventWrappers` array, since the same callback will likely be used for few nodes/events, whereas a node will likely have many events/callbacks.
I've updated the test case to synchronously attach 1000 event listeners and use the same callback for 2 events only (not the same callback for all the events) https://gist.github.com/valdrinkoshi/32d33fd66b7cbcd48a9f70f30fb572ce
Here the results from my tests:
- native (commented the shadydom script): 2 - 3.5ms
- shadydom (master): 3.3 - 6ms (1.65-3x slower than native)
- shadydom (this PR): 3.8 - 6.6ms (1.9-3.3x slower than native)

PTAL
